contents/functionalProgramming/intro.md

```diff
@@ -0,0 +1,117 @@
+
+There are useful functions and techniques that are based off Higher Order Functions in Functional Programming.
+
+- [Currying](https://hackernoon.com/javascript-and-functional-programming-currying-pt-4-96e3230782ab)
+- [Map, Reduce and Filter](https://medium.freecodecamp.org/higher-order-functions-in-javascript-d9101f9cf528) 
+
+## Advantages of Functional Programming
+
+The common question when learning about Functional Programming is what's the point of having us going through all the trouble of having pure functions, state immutability and absence of loops. Turns out, there are various positive implications of Functional Programming:
+
+- [More Bug-Free Code](https://www.quora.com/Are-software-written-using-Functional-Programming-less-buggy-more-robust-and-stable)
+- [Concurrency](https://softwareengineering.stackexchange.com/questions/293851/what-is-it-about-functional-programming-that-makes-it-inherently-adapted-to-para)
+- [And Many More](https://alvinalexander.com/scala/fp-book/benefits-of-functional-programming)
+
+## Disadvantages of Functional Programming
+
+Functional Programming is not a new concept and has been around since the 1950s. Even though it is gaining in popularity today, it is not the predominant programming paradigm used in software applications today. Despite the stated benefits of Functional Programming, there are some downsides of it that can help explain why it is not the mainstream programming paradigm:
+
+- Lack of imperative data structures. One good example is the [lack of hash map](https://stackoverflow.com/questions/6793259/how-does-one-implement-hash-tables-in-a-functional-language) which is an important performant dictionary.
+- Functional Programming can be [slower](https://www.quora.com/Do-functional-programming-languages-always-run-slower-than-imperative-language) than optimal Imperative Programming for reasons such as data copying due to data immutability.
+- And various [other concerns](http://flyingfrogblog.blogspot.sg/2016/05/disadvantages-of-purely-functional.html)
+
+Some of these reasons help explain why 'impure' Functional languages like Scala is much more popular today than Haskell. Scala is a general purpose language that has interoperability with Java and has support for both popular Object Oriented Programming and Functional Programming paradigm. This establishes a [middle ground](https://cacm.acm.org/magazines/2014/4/173220-unifying-functional-and-object-oriented-programming-with-scala/fulltext) for programmers new to Functional Programming.
+
+## Guides on Functional Programming
+
+Functional Programming can be a very different programming paradigm and it definitely takes time to practice to be an expert at it. To learn more about Functional Programming, I would recommend looking at these amazing guides for a deeper understanding:
+
+- [Introduction to Functional Programming](https://medium.com/@cscalfani/so-you-want-to-be-a-functional-programmer-part-1-1f15e387e536)
```

Comment:
I think these are very high quality curated resources, nice! (I haven't took that course myself though just knew that it is very famous 😆 ). I also could even see clearly the different intentions. I'm suggesting that maybe we could even make it even better, something like this:

* An introduction to functional programming, only a three-part series!
* If you are hungry for more, take the excellent functional programming principles course
* And to help with the mindset shift for functional programming, take a look at a practical guide on how to translate an imperative to functional style

Also, I think maybe we should include the [wikipedia](https://en.wikipedia.org/wiki/Functional_programming) page too because this is one of the times it has an excellent overview of functional programming.